### PR TITLE
Allow reading utf-8 encoded json files 

### LINKED ISSUE #1308: Cannot read attributes that contain non-ASCII characters
### Zarr version

2.13.4.dev68

### Numcodecs version

0.11.0

### Python Version

3.10.6

### Operating System

Linux

### Installation

With pip, using the instructions here https://zarr.readthedocs.io/en/stable/contributing.html

### Description

I expect zarr to be able to read attribute files that contain non-ASCII characters, because JSON files use utf-8 encoding. However, currently, there is a check to error if the JSON file contains any non-ASCII characters.

### Steps to reproduce

```python
import zarr
import tempfile
tempdir = tempfile.mkdtemp()
f = open(tempdir + '/.zgroup','w', encoding='utf-8')
f.write('{"zarr_format":2}')
f.close()
f = open(tempdir + '/.zattrs','w', encoding='utf-8')
f.write('{"foo": "た"}')
f.close()
z = zarr.open(tempdir, mode='r')
z.attrs['foo']
```

```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/root/github/zarr-python/zarr/attrs.py", line 74, in __getitem__
    return self.asdict()[item]
  File "/root/github/zarr-python/zarr/attrs.py", line 55, in asdict
    d = self._get_nosync()
  File "/root/github/zarr-python/zarr/attrs.py", line 48, in _get_nosync
    d = self.store._metadata_class.parse_metadata(data)
  File "/root/github/zarr-python/zarr/meta.py", line 104, in parse_metadata
    meta = json_loads(s)
  File "/root/github/zarr-python/zarr/util.py", line 56, in json_loads
    return json.loads(ensure_text(s, 'ascii'))
  File "/root/pyenv/zarr-dev/lib/python3.10/site-packages/numcodecs/compat.py", line 181, in ensure_text
    s = codecs.decode(s, encoding)
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe3 in position 9: ordinal not in range(128)
```

### Additional output

_No response_

### PR DESCRIPTION
**Pull request by nhz2**
Fixes #1308 

Currently, zarr-python errors when reading a json file with non-ascii characters encoded with utf-8, however, Zarr.jl writes json files that include non-ascii characters using utf-8 encoding.
This PR will enable zarr attributes written in Zarr.jl to be read by zarr-python.

TODO:
* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* [x] New/modified features documented in docs/tutorial.rst
* [x] Changes documented in docs/release.rst
* [x] GitHub Actions have all passed
* [x] Test coverage is 100% (Codecov passes)

### CODE REVIEW

**Comment by nhz2** [comment]
@MSanKeys963 @jakirkham  This PR is ready for review. Also, because I am a first-time contributor, I am not able to trigger Github Actions myself.

**Comment by MSanKeys963** [comment]
Thanks for sending the PR @nhz2.
@jakirkham, @joshmoore and @rabernat, please have a look.

**Comment by joshmoore** [comment]
Thanks for this, @nhz2! I've triggered the actions and will go through the code ASAP.

**Comment by joshmoore** [comment]
Green :tada: in case anyone else wants to now take a closer look. (`release.rst` will need its conflict handled)

**Comment by nhz2** [comment]
This PR should have no effect on the JSON files zarr-python writes, they will still be ASCII only with non-ASCII characters escaped. It only affects what kind of JSON files zarr-python can read.

**Comment by joshmoore** [comment]
Thanks, @nhz2. Guess I was more concerned whether any calling code (and/or subclasses) would need modification.

**Comment by joshmoore** [comment]
Thanks for the analysis, @nhz2! Sounds like it should not impact external consumers.

Any objections to rolling this into 2.14, anyone?

File: zarr/meta.py [hunk 1]
```diff
@@ -92,7 +92,7 @@ class Metadata2:
     ZARR_FORMAT = ZARR_FORMAT
 
     @classmethod
-    def parse_metadata(cls, s: Union[MappingType, str]) -> MappingType[str, Any]:
+    def parse_metadata(cls, s: Union[MappingType, bytes]) -> MappingType[str, Any]:
```
Comment (jakirkham) [inline zarr/meta.py:95]: Should this still include `str` as an option? Same question with the functions below

```suggestion
    def parse_metadata(cls, s: Union[MappingType, bytes, str]) -> MappingType[str, Any]:
```

**Comment by jakirkham** [comment]
Thanks for the PR Nathan! 🙏

This is a good improvement to include.

Noting some changes in the typing to still allow `str` in functions (since it should still technically be valid). Though agree `bytes` should be included in a few places as well for clarity

Ideally we would capture any `bytes`-like type here (since they are all valid), but that is somewhat complicated to do in practice. Maybe there will be guidance on this in the future ( https://github.com/python/typing/issues/593 )

**Review by jakirkham — APPROVED** [review]

**Comment by joshmoore** [comment]
I think this has failed on the zarr-feedstock: https://github.com/conda-forge/zarr-feedstock/pull/72

File: zarr/tests/test_attrs.py [hunk 2]
```diff
@@ -42,11 +43,17 @@ def test_storage(self, zarr_version):
...
+        d = json.loads(str(store[attrs_key], 'utf-8'))
         if zarr_version == 3:
             d = d['attributes']
         assert dict(foo='bar', baz=42) == d
 
+    def test_utf8_encoding(self, zarr_version):
+
+        # fixture data
+        fixture = group(store=DirectoryStore('fixture'))
+        assert fixture['utf8attrs'].attrs.asdict() == dict(foo='た')
```
Comment (jakirkham) [inline zarr/tests/test_attrs.py:55]: This test is failing there

File: fixture/utf8attrs/.zattrs [hunk 3]
```diff
@@ -0,0 +1 @@
+{"foo": "た"}
```
Comment (jakirkham) [inline fixture/utf8attrs/.zattrs:1]: This file is missing from the `sdist` on PyPI

File: fixture/utf8attrs/.zgroup [hunk 4]
```diff
@@ -0,0 +1,3 @@
+{
+    "zarr_format": 2
+}
```
Comment (jakirkham) [inline fixture/utf8attrs/.zgroup:3]: Same with this one

**Review by jakirkham — COMMENTED** [review]
The issue is neither of these files were included in the Python `sdist`

File: fixture/utf8attrs/.zattrs (hunk 3 above)
Comment (joshmoore) [reply fixture/utf8attrs/.zattrs:1]: Thanks, @jakirkham. Argh, I thought I had caught these issues but likely only if there's code to generate it them as well. Are you already working on a fix or shall I?

Comment (jakirkham) [reply fixture/utf8attrs/.zattrs:1]: Feel free to go ahead. I might not get to this for a while

Comment (jakirkham) [reply fixture/utf8attrs/.zattrs:1]: Forget if we already discussed this before, but maybe we can make some fixes to CI to ensure we catch these issues in PRs. Wrote up some thoughts on this in issue ( https://github.com/zarr-developers/zarr-python/issues/1347 )

**Pull request merged**